fix(a11y): ARIA nodes should have an accessible name

xlisachan opened the pull request:
## Because

- we are missing ARIA attributes within `<div class="portal" id="dialogs" role="dialog">` DOM elements

## This pull request

- [x] updates `AlertBar` and `DialogMessage` components to add aria-labelledby and aria-describedby attributes
- [x] updates components using `AlertBar` and `DialogMessage` component to include aria-labelledby and aria-describedby attributes
- [x] updates `Portal` component as `modal` is no longer the only id that contain aria attributes
- [x] adds tests to check that aria attributes are also passed when id is set to `dialogs` or `top-bar`

## Issue that this pull request solves

Closes: #12922

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.

## Screenshots
- `id` set to `dialogs`
<img width="1576" alt="Screen Shot 2022-07-05 at 8 28 28 PM" src="https://user-images.githubusercontent.com/28129806/177437952-29476b51-e8e3-4008-a01a-a64cf1998d6f.png">

- `id` set to `modal`
<img width="1579" alt="Screen Shot 2022-07-05 at 8 29 35 PM" src="https://user-images.githubusercontent.com/28129806/177438063-7bb882db-c406-4cec-a0d8-18055d05d137.png">

- `id` set to `top-bar`
<img width="1587" alt="Screen Shot 2022-07-05 at 8 26 37 PM" src="https://user-images.githubusercontent.com/28129806/177437836-aaeda5ea-e496-4fc4-95e0-0ad0b508fcd3.png">

pdehaan commented on packages/fxa-react/components/Portal/index.tsx:
```diff
@@ -45,9 +45,12 @@ const Portal = ({
     }
 
-    if (id === 'modal') {
+    if (id === 'dialogs' || id ==='modal') {
```
nit: add space after `===`
```suggestion
    if (id === 'dialogs' || id === 'modal') {
```

LZoog commented on packages/fxa-payments-server/src/components/DialogMessage/index.stories.tsx:
```diff
@@ -16,7 +16,11 @@ storiesOf('components/DialogMessage', module)
         {({ dialogShown, hideDialog }) =>
           dialogShown && (
-            <DialogMessage onDismiss={hideDialog}>
+            <DialogMessage
```
If you wanna be extra cool you could refactor this to the new storybook format 🙈

xlisachan replied:
I'll keep this in mind and see how far I get with my other two tickets. I know I'll eventually have to refactor this. 🙃

LZoog commented on packages/fxa-react/components/Portal/index.tsx:
```diff
@@ -45,9 +45,12 @@ const Portal = ({
     }
 
-    if (id === 'modal') {
+    if (id === 'dialogs' || id === 'modal') {
```
These are the only possible IDs right? I think you can just take these out of the `if` if so.

xlisachan replied:
I believe there are a total of four: `alert-bar-portal`, `dialogs`, `modal`, and `top-bar`. I'll add `headerId` and `descId` to the other two and then remove the `if`. 👍

LZoog commented on packages/fxa-payments-server/src/App.tsx:
```diff
@@ -180,7 +180,12 @@ export const AppErrorDialog = ({ error: { message } }: { error: Error }) => {
+      <DialogMessage
+        className="dialog-error"
+        onDismiss={locationReload}
+        headerId="general-application-error-header"
```
You'll need to assign these new IDs to the headers/content. See here: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role

pdehaan commented on packages/fxa-payments-server/src/routes/Subscriptions/Reactivate/ConfirmationDialog.tsx:
```diff
@@ -147,20 +157,37 @@ const ConfirmationDialog = ({
+  let ariaLabelledBy;
+  let ariaDescribedBy;
+
+  ariaLabelledBy = !loading && error ? ariaLabelledByError : ariaLabelledByConfirmation;
+  ariaDescribedBy = !loading && !error ? ariaDescribedByError : ariaDescribedByConfirmation;
```
Q: Should we just set these above on L166-167 and change `let` 👉 `const` since the values don't seem to change?

pdehaan commented on packages/fxa-payments-server/src/App.tsx:
```diff
@@ -176,21 +176,23 @@ export class AppErrorBoundary extends React.Component {
 export const AppErrorDialog = ({ error: { message } }: { error: Error }) => {
   const { locationReload } = useContext(AppContext);
+  const ariaLabelledBy = "general-application-error-header";
+  const ariaDescribedBy = "general-application-error-description"
```
```suggestion
  const ariaDescribedBy = "general-application-error-description";
```

pdehaan commented on packages/fxa-payments-server/src/routes/Subscriptions/SubscriptionItem.tsx:
```diff
@@ -69,15 +76,22 @@ export const SubscriptionItem = ({
     !((total || total === 0) && period_start)
   ) {
+    const ariaLabelledBy = "invoice-not-found-header";
+    const ariaDescribedBy = "error-product-plan-not-found-description";
```
Q: Can you confirm this one?
I think this is the only instance I saw that has a different base `ariaLabelledBy` vs `ariaDescribedBy`.
Not sure if this should be `"invoice-not-found-description"`.

xlisachan replied:
Oops, may have mis-copied/pasta'd. Thanks for catching :)

pdehaan commented on packages/fxa-payments-server/src/routes/Checkout/index.tsx:
```diff
@@ -68,16 +68,20 @@ import { CouponDetails } from 'fxa-shared/dto/auth/payments/coupon';
+      headerId={ariaLabelledBy}
+      localizedId="newsletter-signup-error"
+    >
+      You're not signed up for product update emails. You can try again in
```
Q: Should this be a smart quote?

pdehaan commented on packages/fxa-payments-server/src/routes/Subscriptions/PaymentUpdateForm.tsx:
```diff
@@ -209,17 +224,19 @@ export const PaymentUpdateForm = ({
+            descId={alertPendingAriaDescribedBy}
+            localizedId="sub-route-idx-updating"
+          >
+            Updating billing information...
```
Not sure if we have rules around `...` vs `…` (although I guess we previously used `...`, so 🤷) /cc @flodolo

flodolo replied:
Copy should use typographic apostrophes, double quotes, and ellipsis (`’“”…` vs `'""...`).

Note that you can now flag @mozilla/fxa-l10n for questions, which is more scalable than flagging individuals ;-)

pdehaan commented on packages/fxa-payments-server/src/routes/Subscriptions/Reactivate/ConfirmationDialog.tsx:
```diff
@@ -79,7 +89,7 @@ const ConfirmationDialogContent = ({
-          <p>
+          <p id={descId}>
             Your access to {productName} will continue, and your billing cycle
             and payment will stay the same. Your next charge will be
```
Does this line need a trailing `{' '}` (we do below on L101/L111).

sardesam commented on packages/fxa-payments-server/src/components/AlertBar/index.tsx:
```diff
@@ -3,20 +3,65 @@
+            <span
+              data-testid="clear-success-alert"
+              className="close"
+              aria-label="Close modal"
```
Since this is a `span` (non-descriptive element), a `role` attribute will have to be added for the `aria-label` to work as intended.

sardesam commented on packages/fxa-payments-server/src/components/DialogMessage/index.stories.tsx:
```diff
@@ -41,9 +50,12 @@ storiesOf('components/DialogMessage', module)
+        headerId="basic-header"
+        descId="basic-description"
+      >
+        <h4 id="basic-header">This is a basic dialog</h4>
+        <p id="basic-description">Content goes in here.</p>
```
And here (see comment on L24-L25).

sardesam commented on packages/fxa-payments-server/src/components/DialogMessage/index.test.tsx:
```diff
@@ -24,8 +29,14 @@ it('renders as expected', () => {
+      headerId="message-header"
+      descId="message-description"
+    >
+      <div id="message-header" data-testid="children">Message for Mom</div>
+      <p id="message-description">Hi mom</p>
```
Same here (see comment on L21-L22).

sardesam commented on packages/fxa-payments-server/src/components/DialogMessage/index.test.tsx:
```diff
@@ -34,8 +45,13 @@ it('accepts an alternate className', () => {
+      headerId="message-header"
+      descId="message-description"
+    >
+      <div id="message-header" data-testid="children">Message for Mom</div>
+      <p id="message-description">Hi mom</p>
```
Same here (see comment on L21-L22).

sardesam commented on packages/fxa-payments-server/src/components/DialogMessage/index.test.tsx:
```diff
@@ -46,8 +62,9 @@ it('calls onDismiss on click outside', () => {
-    <DialogMessage>
-      <div data-testid="children">Hi mom</div>
+    <DialogMessage headerId="dismiss-header" descId="dismiss-description">
+      <div id="dismiss-header" data-testid="children">Message for Mom</div>
+      <p id="dismiss-description">Hi mom</p>
```
And here (see comment on L21-L22).

sardesam commented on packages/fxa-payments-server/src/components/DialogMessage/index.tsx:
```diff
@@ -34,9 +38,10 @@ export const DialogMessage = ({
+    <Portal id="dialogs" {...{ headerId, descId }}>
       <div data-testid={testid} className={classNames('blocker', 'current')}>
         <div
+          id={headerId}
```
Duplicate `id` to L41.

sardesam commented on packages/fxa-payments-server/src/components/DialogMessage/index.tsx:
```diff
@@ -58,7 +63,7 @@ export const DialogMessage = ({
             </Localized>
           )}
-          <div className="message">{children}</div>
+          <div id={descId} className="message">{children}</div>
```
Duplicate `id` to L41.

sardesam commented on packages/fxa-payments-server/src/components/FetchErrorDialogMessage.tsx:
```diff
@@ -14,10 +17,15 @@ const FetchErrorDialogMessage = ({
+    descId={ariaDescribedBy}
+  >
+    <h4 id={ariaLabelledBy} data-testid={testid}>{title}</h4>
     {fetchState.error && fetchState.error.message && (
-      <p>{fetchState.error.message}</p>
+      <p id={ariaDescribedBy}>{fetchState.error.message}</p>
```
Duplicate `id`s L26 and L23, and L28 and L24.

sardesam commented on packages/fxa-payments-server/src/components/PlanErrorDialog/index.tsx:
```diff
@@ -29,12 +32,16 @@ export const PlanErrorDialog = ({
       <Localized id="product-plan-not-found">
-        <h4>Plan not found</h4>
+        <h4 id={ariaLabelledBy}>Plan not found</h4>
       </Localized>
       <Localized id="product-no-such-plan">
-        <p data-testid="no-such-plan-error">No such plan for this product.</p>
+        <p id={ariaDescribedBy} data-testid="no-such-plan-error">No such plan for this product.</p>
```
Duplicate `id`s L41 and L37, and L44 and L38.

sardesam commented on packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/index.tsx:
```diff
@@ -101,11 +103,19 @@ export const SubscriptionUpgrade = ({
           <Localized id="sub-change-failed">
-            <h4 data-testid="error-plan-update-failed">Plan change failed</h4>
+            <h4
+            id={ariaLabelledBy}
```
Duplicate `id` to L106.

sardesam commented on packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/index.tsx:
```diff
@@ -101,11 +103,19 @@ export const SubscriptionUpgrade = ({
+            Plan change failed</h4>
           </Localized>
-          <p>{updateSubscriptionPlanStatus.error.message}</p>
+          <p id={ariaDescribedBy}>
```
Duplicate `id` to L107.

sardesam commented on packages/fxa-payments-server/src/routes/Subscriptions/PaymentUpdateForm.tsx:
```diff
@@ -231,13 +248,15 @@ export const PaymentUpdateForm = ({
             <img id="error-icon" src={errorIcon} alt="error icon" />
             <Localized id="sub-route-payment-modal-heading">
-              <h2>Invalid billing information</h2>
+              <h2 id={ariaLabelledBy}>Invalid billing information</h2>
```
Duplicate `id` to L251.

sardesam commented on packages/fxa-payments-server/src/routes/Subscriptions/PaymentUpdateForm.tsx:
```diff
@@ -231,13 +248,15 @@ export const PaymentUpdateForm = ({
             </Localized>
             <Localized id="sub-route-payment-modal-message">
-              <p>
+              <p id={ariaDescribedBy}>
```
Duplicate `id` to L252.

sardesam commented on packages/fxa-payments-server/src/routes/Subscriptions/Reactivate/SuccessDialog.tsx:
```diff
@@ -29,8 +32,10 @@ const SuccessDialog = ({
+      descId={ariaDescribedBy}
     >
-      <div className="dialog-icon" style={{ ...setWebIconBackground }}>
+      <div id={ariaLabelledBy} className="dialog-icon" style={{ ...setWebIconBackground }}>
```
Duplicate `id` to L35.

sardesam commented on packages/fxa-payments-server/src/routes/Subscriptions/Reactivate/SuccessDialog.tsx:
```diff
@@ -39,6 +44,7 @@ const SuccessDialog = ({
         />
       </div>
       <p
+        id={ariaDescribedBy}
```
Duplicate `id` to L36.

sardesam commented on packages/fxa-payments-server/src/routes/Subscriptions/SubscriptionItem.tsx:
```diff
@@ -51,15 +51,22 @@ export const SubscriptionItem = ({
         <Localized id="product-plan-not-found">
-          <h4 data-testid="error-subhub-missing-plan">Plan not found</h4>
+          <h4 id={ariaLabelledBy} data-testid="error-subhub-missing-plan">Plan not found</h4>
         </Localized>
         <Localized id="sub-item-no-such-plan">
-          <p>No such plan for this subscription.</p>
+          <p id={ariaDescribedBy}>No such plan for this subscription.</p>
```
Duplicate `id`s (L66 and L62) and (L69 and L63).

sardesam commented on packages/fxa-payments-server/src/routes/Subscriptions/SubscriptionItem.tsx:
```diff
@@ -69,15 +76,22 @@ export const SubscriptionItem = ({
         <Localized id="invoice-not-found">
-          <h4 data-testid="error-subhub-missing-subsequent-invoice">
+          <h4 id={ariaLabelledBy} data-testid="error-subhub-missing-subsequent-invoice">
             Subsequent invoice not found
           </h4>
         </Localized>
         <Localized id="sub-item-no-such-subsequent-invoice">
-          <p>Subsequent invoice not found for this subscription.</p>
+          <p id={ariaDescribedBy}>Subsequent invoice not found for this subscription.</p>
```
Duplicate `id`s (L89 and L85) and (L94 and L86).

sardesam commented on packages/fxa-payments-server/src/routes/Subscriptions/index.tsx:
```diff
@@ -248,42 +253,34 @@ export const Subscriptions = ({
           <Localized id="sub-route-idx-reactivating">
-            <h4 data-testid="error-reactivation">
+            <h4 id={ariaLabelledByReactivateSubscription} data-testid="error-reactivation">
               Reactivating subscription failed
             </h4>
           </Localized>
-          <p>{reactivateSubscriptionStatus.error.message}</p>
+          <p id={ariaDescribedByReactivateSubscription}>
```
Duplicate `id`s to L273-L274.

sardesam commented on packages/fxa-payments-server/src/routes/Subscriptions/index.tsx:
```diff
@@ -298,13 +295,17 @@ export const Subscriptions = ({
           <Localized id="sub-route-idx-cancel-failed">
-            <h4 data-testid="error-cancellation">
+            <h4 id={ariaLabelledByCancellationStatus} data-testid="error-cancellation">
               Cancelling subscription failed
             </h4>
           </Localized>
-          <p>{cancelSubscriptionStatus.error.message}</p>
+          <p id={ariaDescribedByCancellationStatus}>
```
Duplicate `id`s to L298-L299.

sardesam commented on packages/fxa-payments-server/src/routes/Subscriptions/index.tsx:
```diff
@@ -479,12 +480,18 @@ const CancellationDialogMessage = ({
+    >
       <Localized id="sub-route-idx-cancel-msg-title">
-        <h4 data-testid="cancellation-message-title">
+        <h4 id={ariaLabelledBy} data-testid="cancellation-message-title">
```
Duplicate `id` to L490.

sardesam commented on packages/fxa-payments-server/src/routes/Subscriptions/index.tsx:
```diff
@@ -495,7 +502,7 @@ const CancellationDialogMessage = ({
         }}
       >
-        <p>
+        <p id={ariaDescribedBy}>
```
Duplicate `id` to L491.

sardesam commented on packages/fxa-react/components/Portal/index.test.tsx:
```diff
@@ -56,7 +56,8 @@ it('applies a11y improvements when id is set to "modal"', () => {
   render(
     <Portal id="modal" {...{ headerId, descId }}>
-      <div>Hi mom</div>
+      <div id={headerId}>Message for mom</div>
+      <p id={descId}>Hi mom</p>
```
Duplicate `id`s to L58.

sardesam commented on packages/fxa-react/components/Portal/index.test.tsx:
```diff
@@ -77,3 +78,51 @@ it('applies a11y improvements when id is set to "modal"', () => {
+
+  render(
+    <Portal id="dialogs" {...{ headerId, descId }}>
+      <div id={headerId}>Message for mom</div>
+      <p id={descId}>Hi mom</p>
```
Duplicate `id`s to L93.

sardesam commented on packages/fxa-react/components/Portal/index.test.tsx:
```diff
@@ -77,3 +78,51 @@ it('applies a11y improvements when id is set to "modal"', () => {
+
+  render(
+    <Portal id="top-bar" {...{ headerId, descId }}>
+      <div id={headerId}>Message for mom</div>
+      <p id={descId}>Hi mom</p>
```
Duplicate `id`s to L117.

sardesam reviewed (changes requested):
Thank you for updating the `aria` attributes and adding additional tests! That makes me so happy 💟 !

The only major change that applies to most of the files is duplication of `id`s for the label and description. They are used for both the modal (content box) and inner copy. `Id`s are unique and cannot be shared.

If it's not being used elsewhere (I checked and the `id` is not attached to styling, but I'm not sure if it's being used in the functionality of the modal at all), the modal parent `id`s should be changed while the copy should remain the same so screen readers know what is being referenced.

Let me know if you have any questions or if I missed anything. Happy to jump on a call or Slack! 🙂

xlisachan commented:
Thanks for the feedback, @sardesam ! Per our conversation, please see changes in the latest commit outlined below - hopefully this makes it easier for you to re-review. :)

(1) `fxa-payments-server/src/components/AlertBar/index.stories.tsx`
  - removed `descId` as `aria-describedby` is not needed in the `AlertBar` component

(2)`fxa-payments-server/src/components/AlertBar/index.test.tsx`
  - removed `descId` as `aria-describedby` is not needed in the `AlertBar` component
  - added tests to check for for `aria-labelledby` and `role` is set to "dialog"
  - updated the id so that it is unique

(3) `fxa-payments-server/src/components/AlertBar/index.tsx`
  - updated so that `headerId` and `descId` are no longer passed into `Portal` which is causing the duplication
  - corrected the id from `descId` to `headerId` on line 44
  - added `role` of button to `span`

(4) `fxa-payments-server/components/DialogMessage/index.stories.tsx`
  - updated the id so that it is unique

(5) `fxa-payments-server/components/DialogMessage/index.test.tsx`
  - updated the id so that it is unique
  - added tests to check for for `aria-labelledby`, `aria-describedby`, and `role` is set to "dialog"

(6)  `fxa-payments-server/components/DialogMessage/index.tsx`
  - updated so that `headerId` and `descId` are no longer passed into `Portal` which is causing the duplication
  - set `aria-labelledby`, `aria-describedby`, and `role` in the correct parent div

(7) `routes/Checkout/index.tsx`
  - removed `descId` as `aria-describedby` is not needed in the `AlertBar` component

(8) `routes/Subscription/PaymentUpdateForm.tsx`
  - removed `descId` as `aria-describedby` is not needed in the `AlertBar` component

(9) `routes/Subscription/index.tsx`
 - removed `descId` as `aria-describedby` is not needed in the `AlertBar` component
 
(10) `fxa-react/components/Portal/index.test.tsx`
  - removed tests that check for aria attributes as these attributes are no longer passed into the `Portal` component
 
(11) `fxa-react/components/Portal/index.tsx`
  - removed headerId and descId from being passed into this component
  - removed setting `role` to dialog
  
(12) `fxa-settings/src/components/Modal/index.test.tsx`
  - updated the id so that it is unique
  - added tests to check for for `aria-labelledby`, `aria-describedby`, and `role` is set to "dialog"
  
(13) `fxa-settings/src/components/Modal/index.tsx`
  - updated so that `headerId` and `descId` are no longer passed into `Portal` which is causing the duplication
  - set `aria-labelledby`, `aria-describedby`, and `role` in the correct parent div

sardesam reviewed (approved):
LGTM! Thanks for making these changes. 😄
